ocaml/xapi/helpers.ml

```diff
@@ -1768,39 +1773,51 @@ end = struct
 
   let systemctl_ cmd = systemctl cmd |> ignore
 
-  let is_enabled () =
-    let is_enabled_stdout =
-      try systemctl "is-enabled"
-      with
-      (* systemctl is-enabled appears to return error code 1 when the service is disabled *)
-      | Forkhelpers.Spawn_internal_error (stderr, stdout, status) as e -> (
-        match status with
-        | Unix.WEXITED n
-          when n = 1
-               && Astring.String.is_prefix ~affix:"disabled" stdout
-               && Astring.String.is_empty stderr ->
-            "disabled"
-        | _ ->
-            raise e
-      )
-    in
-    is_enabled_stdout |> Astring.String.trim |> function
-    | "enabled" ->
-        true
-    | "disabled" ->
-        false
-    | unknown ->
-        D.error
-          "Stunnel.is_enabled: expected 'enabled' or 'disabled', but got: %s"
-          unknown ;
-        false
+  let systemctl_socket cmd =
+    call_script !Xapi_globs.systemctl [cmd; "stunnel-on-demand.socket"]
+
+  let systemctl_socket_ cmd = systemctl_socket cmd |> ignore
+
+  let systemctl_socket_stop_on_demand_ () =
+    call_script !Xapi_globs.systemctl ["stop"; "stunnel-on-demand@*.service"]
+    |> ignore
+
+  let _enable () =
+    if is_fips_enabled () then (
+      systemctl_ "stop" ; systemctl_ "disable" ; systemctl_socket_ "enable"
+    ) else (
+      systemctl_socket_ "stop" ;
+      (* This will stop all stunnel-on-demand@*.service instances *)
+      systemctl_socket_ "disable" ;
+      systemctl_ "enable"
+    )
+
+  let _restart () =
+    if is_fips_enabled () then
+      (* This may fail with 'failed to listen on sockets: Address already in use'
+       * in restart after pool.eject.
+       *)
+      retry_until_timeout ~interval:0.2 ~timeout:30.
+        "restarting stunnel-on-demand.socket" (fun () ->
+          try
+            (* To make all existing instances be terminated before restarting,
+             * the systemd socket needs to be stopped and then started.
+             *)
+            systemctl_socket_ "stop" ;
+            systemctl_socket_stop_on_demand_ () ;
+            systemctl_socket_ "start" ;
+            true
+          with e -> false)
```

Comment:
If something goes wrong, create a log entry.

Reply:
The functions will call `call_script` eventually in which it will write log.